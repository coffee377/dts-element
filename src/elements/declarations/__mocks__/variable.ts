export class VariableDeclaration {

  public parameters: any;

  constructor(parameters: any) {
    this.parameters = parameters;
  }

  public _emit(): string {
    return `[VariableDeclaration ${this.parameters.name}]`;
  }

}
