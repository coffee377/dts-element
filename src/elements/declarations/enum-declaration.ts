import {emit_declare} from '../../helpers/emit-declare';
import {emit_enum_members} from '../../helpers/emit-enum-members';
import {Stack} from '../../stack';
import {Declaration, IDeclarationOptionalParameters} from '../declaration';
import {EnumMember} from '../members/enum-member';

export interface IEnumDeclarationRequiredParameters {
  name: string;
}

export interface IEnumDeclarationOptionalParameters {
  members: EnumMember[];
}

export class EnumDeclaration
    extends Declaration<IEnumDeclarationRequiredParameters, IEnumDeclarationOptionalParameters> {

  public get default_parameters(): IDeclarationOptionalParameters & IEnumDeclarationOptionalParameters {
    return Object.assign({}, super.default_declaration_parameters, {
      members: [],
    });
  }

  public _emit_raw(stack: Stack): string {
    const {name, members} = this.parameters;
    const declare = emit_declare(stack);
    return `${declare}enum ${name} ${emit_enum_members(members, stack)}`;
  }

}
