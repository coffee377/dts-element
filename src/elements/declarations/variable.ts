import {any_type} from '../../constants';
import {AnyElement} from '../../element';
import {Declaration, IDeclarationOptionalParameters} from '../declaration';
import {AnyType} from '../type';

// tslint:disable-next-line no-empty-interface
export interface IVariableDeclarationRequiredParameters {}

export interface IVariableDeclarationOptionalParameters {
  kind: 'var' | 'let' | 'const';
  type: AnyType;
  optional: boolean;
}

export class VariableDeclaration
    extends Declaration<IVariableDeclarationRequiredParameters, IVariableDeclarationOptionalParameters> {

  public _emit(_container: AnyElement | null): string {
    const {name, kind, type: a_type, optional} = this.parameters;
    const op = ('?').repeat(+optional);
    return `declare ${kind} ${name}${op}: ${a_type._emit(this)};`;
  }

  public get default_parameters(): IDeclarationOptionalParameters & IVariableDeclarationOptionalParameters {
    return Object.assign({}, super.default_declaration_parameters, {
      kind: 'var' as 'var',
      type: any_type,
      optional: false,
    });
  }

}
