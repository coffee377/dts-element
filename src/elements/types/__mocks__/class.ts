import {Type} from '../../type';

export class ClassType extends Type<any, any> {

  public get default_parameters(): any {
    return {};
  }

  public _emit(): string {
    return `[ClassType ${this.parameters.class}]`;
  }

}
