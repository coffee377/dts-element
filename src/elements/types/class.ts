import {Container} from '../../collections';
import {emit_generics} from '../../helpers/emit-generics';
import {Type} from '../type';

// TODO: support for ClassDeclaration
export interface IClassTypeRequiredParameters {
  class: string;
}

export interface IClassTypeOptionalParameters {
  generics: Type[];
}

export class ClassType extends Type<IClassTypeRequiredParameters, IClassTypeOptionalParameters> {

  public get default_parameters(): IClassTypeOptionalParameters {
    return {
      generics: [],
    };
  }

  public _emit(_container: Container): string {
    const name = this.parameters.class;
    return `${name}${emit_generics(this.parameters.generics, this)}`;
  }

}
