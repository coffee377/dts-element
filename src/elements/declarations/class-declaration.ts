import {Container} from '../../collections';
import {emit_abstract} from '../../helpers/emit-abstract';
import {emit_extends} from '../../helpers/emit-extends';
import {emit_generics} from '../../helpers/emit-generics';
import {emit_members} from '../../helpers/emit-members';
import {Declaration, IDeclarationOptionalParameters} from '../declaration';
import {ClassMember} from '../members/class-member';
import {ClassType} from '../types/class-type';
import {GenericType} from '../types/generic-type';

export interface IClassDeclarationRequiredParameters {
  name: string;
}

export interface IClassDeclarationOptionalParameters {
  generics: GenericType[];
  members: ClassMember[];
  extends: ClassType | null;
  abstract: boolean;
}

export class ClassDeclaration
    extends Declaration<IClassDeclarationRequiredParameters, IClassDeclarationOptionalParameters> {

  public get default_parameters(): IDeclarationOptionalParameters & IClassDeclarationOptionalParameters {
    return Object.assign({}, super.default_declaration_parameters, {
      generics: [],
      members: [],
      extends: null,
      abstract: false,
    });
  }

  public _emit_raw(container: Container): string {
    const {name, members, generics} = this.parameters;
    const generic = emit_generics(generics, this);
    const abstract = emit_abstract(this.parameters.abstract, this);
    const an_extends = emit_extends(this.parameters.extends, this);
    return `${abstract}class ${name}${generic}${an_extends} ${emit_members(members, this)}`;
  }

}
