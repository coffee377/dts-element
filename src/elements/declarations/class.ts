import {Container} from '../../collections';
import {emit_abstract} from '../../helpers/emit-abstract';
import {emit_declaration_name} from '../../helpers/emit-declaration-name';
import {emit_extends} from '../../helpers/emit-extends';
import {emit_generics} from '../../helpers/emit-generics';
import {emit_object} from '../../helpers/emit-object';
import {Declaration, IDeclarationOptionalParameters} from '../declaration';
import {ClassMember} from '../members/class';
import {ClassType} from '../types/class';
import {GenericType} from '../types/generic';

export interface IClassDeclarationRequiredParameters {
  name: string;
}

export interface IClassDeclarationOptionalParameters {
  generics: GenericType[];
  children: ClassMember[];
  extends: ClassType | null;
  abstract: boolean;
}

export class ClassDeclaration
    extends Declaration<IClassDeclarationRequiredParameters, IClassDeclarationOptionalParameters> {

  public get default_parameters(): IDeclarationOptionalParameters & IClassDeclarationOptionalParameters {
    return Object.assign({}, super.default_declaration_parameters, {
      generics: [],
      children: [],
      extends: null,
      abstract: false,
    });
  }

  public _emit_raw(container: Container): string {
    const {children, generics} = this.parameters;
    const name = emit_declaration_name(this.parameters.name, container);
    const generic = emit_generics(generics, this);
    const abstract = emit_abstract(this.parameters.abstract, this);
    const an_extends = emit_extends(this.parameters.extends, this);
    return `${abstract}class ${name}${generic}${an_extends} ${emit_object(children, this)}`;
  }

}
