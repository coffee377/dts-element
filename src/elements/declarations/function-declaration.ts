import {any_type} from '../../constants';
import {emit_declaration_name} from '../../helpers/emit-declaration-name';
import {emit_declare} from '../../helpers/emit-declare';
import {emit_export} from '../../helpers/emit-export';
import {emit_function} from '../../helpers/emit-function';
import {emit_generics} from '../../helpers/emit-generics';
import {emit_parameters} from '../../helpers/emit-parameters';
import {Stack} from '../../stack';
import {Declaration, IDeclarationOptionalParameters} from '../declaration';
import {Parameter} from '../parameter';
import {Type} from '../type';
import {TypeAssertion} from '../type-assertion';
import {FunctionType} from '../types/function-type';
import {GenericType} from '../types/generic-type';

// tslint:disable-next-line no-empty-interface
export interface IFunctionDeclarationRequiredParameters {}

export interface IFunctionDeclarationOptionalParameters {
  type: FunctionType;
  export: boolean;
}

export class FunctionDeclaration
    extends Declaration<IFunctionDeclarationRequiredParameters, IFunctionDeclarationOptionalParameters> {

  private _instance_of_function_declaration: true;

  public get default_parameters(): IDeclarationOptionalParameters & IFunctionDeclarationOptionalParameters {
    return Object.assign({}, super.default_declaration_parameters, {
      type: new FunctionType({return: any_type}),
      export: false,
    });
  }

  public _emit_raw(stack: Stack): string {
    const {type} = this.parameters;
    const name = emit_declaration_name(this.parameters.name, stack);
    const parameters = emit_parameters(type.parameters.parameters, stack);
    const generics = emit_generics(type.parameters.generics, stack, true);
    const return_type = type.parameters.return.emit(stack);
    const declare = emit_declare(stack);
    const func = emit_function(stack);
    const an_export = emit_export(this.parameters.export, stack);
    return `${an_export}${declare}${func}${name}${generics}(${parameters}): ${return_type};`;
  }

}
