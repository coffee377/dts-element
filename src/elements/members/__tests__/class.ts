jest.unmock('../class');

import {VariableDeclaration} from '../../declarations/variable';
import {ClassMember} from '../class';

describe('#emit()', () => {
  it('should return correctly', () => {
    expect(new ClassMember({
      owned: new VariableDeclaration({name: 'x'}),
    }).emit()).toMatchSnapshot();
  });
});
