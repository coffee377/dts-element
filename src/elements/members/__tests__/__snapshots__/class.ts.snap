// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#emit() should return correctly 1`] = `"[emit-jsdoc][emit-accessor][emit-static][emit-abstract][emit-readonly][VariableDeclaration x]"`;
